### PR TITLE
Bugfix: DemuxFastqs should set the pf/non-pf metrics correctly after #713

### PR DESCRIPTION
**Pull request by nh13**

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 1]
```diff
@@ -810,7 +809,11 @@ private class FastqDemultiplexer(val sampleInfos: Seq[SampleInfo],
 
     val passQc = demuxRecords.forall(d => d.readInfo.forall(_.passQc))
     val isControl = demuxRecords.forall(d => d.readInfo.forall(_.internalControl))
-    DemuxResult(sampleInfo=sampleInfo, numMismatches=numMismatches, records=demuxRecords, passQc=passQc, isControl=isControl)
+
+    val result = DemuxResult(sampleInfo=sampleInfo, numMismatches=numMismatches, records=demuxRecords, passQc=passQc, isControl=isControl)
+    result.sampleInfo.metric.increment(numMismatches = numMismatches, isPf = passQc)
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:814]:
```suggestion
    result.sampleInfo.metric.increment(numMismatches=numMismatches, isPf=passQc)
```

File: src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala [hunk 2]
```diff
@@ -964,6 +964,60 @@ class DemuxFastqsTest extends UnitSpec with OptionValues with ErrorLogLevel {
...
 
+
+  def testEndToEndBarcodeMetrics(fastqStandards: FastqStandards, omitFailingReads: Boolean = false, omitControlReads: Boolean = false): Unit = {
+    // Build the FASTQ
+    val fastqs = new ListBuffer[FastqRecord]()
+    val namePrefix = "Instrument:RunID:FlowCellID:Lane:Tile:X"
+    val qualities = "?"*17 + Range.inclusive(2, 40).map(q => (q + 33).toChar).mkString
+    fastqs += fq(name=f"$namePrefix:1", comment=Some("1:Y:0:SampleNumber"), bases=sampleBarcode1 + "A"*39, quals=Some(qualities)) // matches the first sample -> first sample
+    fastqs += fq(name=f"$namePrefix:2", comment=Some("2:N:0:SampleNumber"), bases=sampleBarcode1 + "G"*39, quals = Some(qualities)) // matches the first sample -> first sample
+//    fastqs += fq(name=f"$namePrefix:2", comment=Some("2:Y:0:SampleNumber"), bases=sampleBarcode2 + "G"*39, quals = Some(qualities)) // matches the second sample -> second sample
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:975]: remove?

File: src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala [hunk 3]
```diff
@@ -964,6 +964,60 @@ class DemuxFastqsTest extends UnitSpec with OptionValues with ErrorLogLevel {
...
+    val metricsFilename = "demux_barcode_metrics.txt"
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:976]: Please make a temporary file, and then specify it to `DemuxFastqs`
```suggestion
    val metricsFilename = makeTempFile("demux_barcode_metrics", ".txt")
```

File: src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala [hunk 4]
```diff
@@ -964,6 +964,60 @@ class DemuxFastqsTest extends UnitSpec with OptionValues with ErrorLogLevel {
...
+  }
+
+  "DemuxFastqs" should "only increment the pf fields for passing reads in the sample barcode metrics" in {
+    testEndToEndBarcodeMetrics(FastqStandards(includeReadNumbers=true, includeSampleBarcodes=true), omitFailingReads = true)
+  }
+
+  it should "only increment the pf fields for passing reads even when --omit-failing-reads is false" in {
+    testEndToEndBarcodeMetrics(FastqStandards(includeReadNumbers=true, includeSampleBarcodes=true), omitFailingReads = false)
+  }
+
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:1018]:
```suggestion
```

File: src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala [hunk 5]
```diff
@@ -58,4 +58,24 @@ class SampleBarcodeMetricTest extends UnitSpec {
     noMatch.pf_ratio_this_barcode_to_best_barcode shouldBe 1d
     noMatch.pf_normalized_matches shouldBe 12/7d +- 0.00001
   }
+
+  "SampleBarcodeMetric.increment" should "increment certain variables when the read doesn't pass QC" in {
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala:62]:
```suggestion
  "SampleBarcodeMetric.increment" should "not increment `pf_*` variables when the read doesn't pass QC" in {
```

File: src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala [hunk 6]
```diff
@@ -58,4 +58,24 @@ class SampleBarcodeMetricTest extends UnitSpec {
...
+    val metric = new SampleBarcodeMetric()
+
+    metric.increment(numMismatches = 1, isPf = false)
+    metric.templates shouldEqual 1
+    metric.pf_templates shouldEqual 0
+    metric.one_mismatch_matches shouldEqual 1
+    metric.pf_one_mismatch_matches shouldEqual 0
+  }
+
+  it should "increment pf variables as well when the read passes QC" in {
```
Comment (nh13) [inline src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala:72]:
```suggestion
  it should "increment `pf_*` variables as well when the read passes QC" in {
```

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 7]
```diff
@@ -810,7 +809,11 @@ private class FastqDemultiplexer(val sampleInfos: Seq[SampleInfo],
 
     val passQc = demuxRecords.forall(d => d.readInfo.forall(_.passQc))
     val isControl = demuxRecords.forall(d => d.readInfo.forall(_.internalControl))
-    DemuxResult(sampleInfo=sampleInfo, numMismatches=numMismatches, records=demuxRecords, passQc=passQc, isControl=isControl)
+
+    val result = DemuxResult(sampleInfo=sampleInfo, numMismatches=numMismatches, records=demuxRecords, passQc=passQc, isControl=isControl)
+    result.sampleInfo.metric.increment(numMismatches=numMismatches, isPf=passQc)
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:814]: Is this the right place for this?  Suppose I am omitting control reads, then they will be counted here.  I think we don't want them counted.  I think you want to increment in `demultiplexingIterator`?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 8]
```diff
@@ -167,7 +167,11 @@ object DemuxFastqs {
       zippedIterator.map { readRecords => demultiplexer.demultiplex(readRecords: _*) }
     }
 
-    resultIterator.filter(r => r.keep(omitFailingReads, omitControlReads))
+    resultIterator.map { res =>
+      val considerRead = if (!omitControlReads) true else !res.isControl
+      if (considerRead) res.sampleInfo.metric.increment(numMismatches = res.numMismatches, isPf = res.passQc)
```
Comment (jacarey) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:172]:
```suggestion
      if (!omitControlReads || !res.isControl) res.sampleInfo.metric.increment(numMismatches = res.numMismatches, isPf = res.passQc)
```

File: src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala [hunk 9]
```diff
@@ -34,28 +34,48 @@ import scala.collection.mutable
 class SampleBarcodeMetricTest extends UnitSpec {
 
   "SampleBarcodeMetric.finalizeMetrics" should "fill out all metrics that require summary counts" in {
-    val metricOne = new SampleBarcodeMetric("Alice", "Alice.lib", "AAAAAAA", 10, 5,   9,  4, 1, 1)
-    val metricTwo = new SampleBarcodeMetric("Bob",   "Bob.lib",   "CCCCCCC", 20, 10, 19,  9, 1, 1)
-    val noMatch   = new SampleBarcodeMetric("Eve",   "Eve.lib",   "NNNNNNN", 30, 20, 29, 19, 1, 1)
+    val metricOne = new SampleBarcodeMetric("Alice", "Alice.lib", "AAAAAAA", 10, 5, 9, 4, 1, 1)
+    val metricTwo = new SampleBarcodeMetric("Bob", "Bob.lib", "CCCCCCC", 20, 10, 19, 9, 1, 1)
+    val noMatch = new SampleBarcodeMetric("Eve", "Eve.lib", "NNNNNNN", 30, 20, 29, 19, 1, 1)
```
Comment (jacarey) [inline src/test/scala/com/fulcrumgenomics/util/SampleBarcodeMetricTest.scala:39]:
```suggestion
    val metricOne = new SampleBarcodeMetric("Alice", "Alice.lib", "AAAAAAA", 10, 5, 9, 4, 1, 1)
    val metricTwo = new SampleBarcodeMetric("Bob", "Bob.lib", "CCCCCCC", 20, 10, 19, 9, 1, 1)
    val noMatch   = new SampleBarcodeMetric("Eve", "Eve.lib", "NNNNNNN", 30, 20, 29, 19, 1, 1)
```

**Review by jacarey — APPROVED** [review]
Minor changes.

**Pull request merged**